File: Leetcode/Tree/102.binary-tree-level-order-traversal.java
/*
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree [3,9,20,null,null,15,7],

    3
   / \
  9  20
    /  \
   15   7

return its level order traversal as:

[
  [3],
  [9,20],
  [15,7]
]
****************************************************************************Solution*************************************************************************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> nodes = new ArrayList<>();
        
        Queue<TreeNode> q = new LinkedList<>();
        
        if(root != null)
            q.add(root);
        
        while(!q.isEmpty())
        {
            int size = q.size();
            
            List<Integer> nodesAtLevel = new ArrayList<>();  //this will hold the nodes placed in same level
            
            while(size-- > 0)
            {                                
                TreeNode currentNode = q.poll();                
                
                nodesAtLevel.add(currentNode.val);
                
                if(currentNode.left != null)
                    q.add(currentNode.left);
                
                if(currentNode.right != null)
                    q.add(currentNode.right);
            }
            
            nodes.add(nodesAtLevel);
        }
        
        return nodes;
    }
}

File: DSA-JavaScript/Tree/102.binary-tree-level-order-traversal.js
/*

Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).


Example 1:

Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]
Example 2:

Input: root = [1]
Output: [[1]]
Example 3:

Input: root = []
Output: []
 

Constraints:

The number of nodes in the tree is in the range [0, 2000].
-1000 <= Node.val <= 1000

****************************************************************Solution********************************************************/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {        
    const queue = [], res = [];
    
    if(!root){
        return res;
    }
    
    queue.push(root);
    
    while(queue.length){
        const nodesOfCurrentLevel = [];
        let widthOfCurrentLevel = queue.length;
        
        while(widthOfCurrentLevel-- > 0){
            const removedNode = queue.shift();
            nodesOfCurrentLevel.push(removedNode.val);
            
            if(removedNode.left){
                queue.push(removedNode.left)
            }
            
            if(removedNode.right){
                queue.push(removedNode.right);
            }
        }
        
        res.push(nodesOfCurrentLevel);
    }
    
    return res;
};